enhancements/sig-architecture/105-aws-iam-registration/README.md

```diff
@@ -0,0 +1,464 @@
+
+The role name and policies created on managed cluster IAM will follow a templates to ensure standardization. The role name will contain md5 hash of the values of tags to uniquely identify.
+
+```shell
+arn:aws:iam::<managed-cluster-account-id>:role/ocm-spoke-<md5-hash-unique-identifier>
+```
+
+Another reason to follow a standard name is to ensure that the role can be created before running `clusteradm join` command and the klusterlet/registration agent can assume that role, on startup, with a standardized name.
+
+#### Graceful resource cleanup on both hub and managed cluster when managed cluster triggers unjoin
+
+The `ManagedCluster` will have a new boolean field `clientUnjoinsHub` with the default value of false. The `clusteradm unjoin` command by managed cluster admin will update the `ManagedCluster` on hub and set the `spec.clientUnjoinsHub` to true, which will trigger the new `aws-irsa` controller on hub to clean-up all the managed cluster specific resources, like IAM roles and policies on hub, and ultimately delete the `ManagedCluster`.
+
+or
+
+The unjoin command will update the `ManagedCluster` on hub and add a new condition `ManagedClusterUnjoined`, which will trigger the new `aws-irsa` controller on hub to clean-up all the managed cluster specific resources, like IAM roles and policies on hub, and ultimately delete the `ManagedCluster`.
+
+
+#### Managed cluster permission isolation on hub
+
+Every managed cluster will have a separate role on both managed cluster as well as hub IAM. Every managed cluster role on hub will allow access to managed cluster's namespace only. All roles will follow a standard naming convention, and the IAM trust policies allow roles to be assumed by relevant resources (pods, clusters, etc) only.
+
+This will ensure that roles can only be assumed by the consumers they are meant for, such that one managed cluster cannot assume roles and access resources belonging to other managed cluster
+
+### Workflow Details
+
+This is to describe the various process of cluster join
+
+Actors:
+1. cluster-admin on managed cluster
+2. AWS IAM admin on managed cluster AWS account
+3. cluster-admin on hub cluster
+4. AWS IAM admin on hub cluster AWS account
+5. hub controller
+6. agent on managed cluster
+
+Some rules on cluster join:
+- The name of the cluster must be globally unique on hub and conforms to dns label format.
+
+#### Managed Cluster Prerequisites
+
+An IAM role named using following template:
+```
+arn:aws:iam::<managed-cluster-account-id>:role/ocm-managed-cluster-<md5-hash-unique-identifier>
+```
+
+Having following permissons access policy:
+```json
+{
+   "Version": "2012-10-17",
+   "Statement": [
+      {
+         "Effect": "Allow",
+         "Action": "sts:AssumeRole",
+         "Resource": "arn:aws:iam::<hub-account-id>:role/ocm-hub-<md5-hash-unique-identifier>"
+      }
+   ]
+}
+```
+
+And following trust relationship policy:
+```json
+{
```

Comment:
If my udnerstanding is correct, <klusterlet-name>-registration-sa is the service account on [registration-agent](https://github.com/open-cluster-management-io/ocm/blob/main/manifests/klusterlet/management/klusterlet-registration-deployment.yaml), when deployed in [non-singleton](https://github.com/open-cluster-management-io/ocm/blob/main/pkg/operator/operators/klusterlet/controllers/klusterletcontroller/klusterlet_runtime_reconcile.go#L75) mode.

The registration-agent will create hub-kubeconfig-secret in managed cluster when the authoization is granted on hub. The hub-kubeconfig generation needs the iam permissions on hub as mentioned [here](https://github.com/suvaanshkumar/enhancements/blob/OCM-using-AWS-IAM/enhancements/sig-architecture/105-aws-iam-registration/README.md?plain=1#L352), that is why we are trying to add this role.

Reply:
my understanding is hub-kubeconfig does not relates to registration-sa. registration-sa is for agent to talk to spoke cluster. It should not be related to the secret that agent talks to the hub.

Reply:
My understanding is same as well. In EKS IRSA roles are assigned through service accounts, we want to use this service account to assign spoke a role on its IAM, which it will use to assume role on hub.

Spoke agent won't assume a role on hub directly, it first assumes a role on spoke's IAM, and using that role assumes role on hub. Please refer "Example Use chained AssumeRole operations" section on [this](https://docs.aws.amazon.com/eks/latest/userguide/cross-account-access.html) article.